contracts/router/quoter/PoolQuoterV1.sol

```diff
@@ -0,0 +1,398 @@
+    {
+        return IDefaultExtendedPool(pool).calculateRemoveLiquidity(amount);
+    }
+
+    /// @inheritdoc ISwapQuoterV1
+    function calculateWithdrawOneToken(
+        address pool,
+        uint256 tokenAmount,
+        uint8 tokenIndex
+    ) external view returns (uint256 amountOut) {
+        return IDefaultExtendedPool(pool).calculateRemoveLiquidityOneToken(tokenAmount, tokenIndex);
+    }
+
+    // ══════════════════════════════════════════════ POOL GETTERS V1 ══════════════════════════════════════════════════
+
+    /// @inheritdoc ISwapQuoterV1
+    function poolInfo(address pool) external view returns (uint256 numTokens, address lpToken) {
+        numTokens = _numTokens(pool);
+        lpToken = _lpToken(pool);
+    }
+
+    /// @inheritdoc ISwapQuoterV1
+    function poolTokens(address pool) external view returns (PoolToken[] memory tokens) {
+        tokens = _getPoolTokens(pool);
+    }
+
+    // ══════════════════════════════════════════════ POOL INSPECTION ══════════════════════════════════════════════════
+
+    /// @dev Returns the LP token address for the given pool, if it exists. Otherwise, returns address(0).
+    function _lpToken(address pool) internal view returns (address) {
+        // Try getting the LP token address from the pool.
+        try IDefaultExtendedPool(pool).swapStorage() returns (
+            uint256,
+            uint256,
+            uint256,
+            uint256,
+            uint256,
+            uint256,
+            address lpToken
+        ) {
+            return lpToken;
+        } catch {
+            // Return address(0) if the pool doesn't have an LP token.
+            return address(0);
+        }
+    }
+
+    /// @dev Returns the number of tokens the given pool supports.
+    function _numTokens(address pool) internal view returns (uint256 numTokens) {
+        while (true) {
+            // Iterate over the tokens until we get an exception.
+            try IDefaultExtendedPool(pool).getToken(uint8(numTokens)) returns (address) {
+                unchecked {
+                    // unchecked: ++numTokens never overflows uint256
+                    ++numTokens;
+                }
+            } catch {
+                // End of pool reached, exit the loop.
+                break;
+            }
+        }
+    }
+
+    /// @dev Returns the tokens the given pool supports.
+    function _getPoolTokens(address pool) internal view returns (PoolToken[] memory tokens) {
+        uint256 numTokens = _numTokens(pool);
+        tokens = new PoolToken[](numTokens);
+        unchecked {
+            // unchecked: ++i never overflows uint256
+            for (uint256 i = 0; i < numTokens; ++i) {
+                address token = IDefaultExtendedPool(pool).getToken(uint8(i));
+                tokens[i] = PoolToken({isWeth: token == _weth, token: token});
+            }
+        }
+    }
+
+    /// @dev Returns pool indexes for the two given tokens plus 1.
```

Comment:
Why +1 here?

Reply:
+1 is used here to assign return value of zero to tokens that are not in the pool. Otherwise you'd need to pass index and a bool flag.

I guess though a return value of 255 could be used for tokens that are not in the pool. Will evaluate whether this might make the code cleaner.

Reply:
@mikeyrf Good call, removed +1 in 77904b10aecad70ccc8cbfcac9e6b39ba414b522

Reply:
Sg!

Reply:
@ChiTimesChi just a heads up on stale comment: `// @dev` comment still says +1

Reply:
Fixed in 10f5e8a8968ad3fb1e96d6893ecdb653e79f8a3b